diffusion/models/autoencoder.py

```diff
@@ -0,0 +1,613 @@
+        for param in self.lpips.net.parameters():
+            param.requires_grad_(False)
+
+        # Set up the discriminator
+        self.discriminator_num_filters = discriminator_num_filters
+        self.discriminator_num_layers = discriminator_num_layers
+        self.discriminator_weight = discriminator_weight
+        self.discriminator = NlayerDiscriminator(input_channels=self.output_channels,
+                                                 num_filters=self.discriminator_num_filters,
+                                                 num_layers=self.discriminator_num_layers)
+        self.scale_gradients = GradientScalingLayer()
+        self.scale_gradients.register_full_backward_hook(self.scale_gradients.backward_hook)
+
+    def set_discriminator_weight(self, weight: float):
+        self.discriminator_weight = weight
+
+    def calc_discriminator_adaptive_weight(self, nll_loss, fake_loss, last_layer):
+        # Need to ensure the grad scale from the discriminator back to 1.0 to get the right norm
+        self.scale_gradients.set_scale(1.0)
+        # Get the grad norm from the nll loss
+        nll_grads = torch.autograd.grad(nll_loss, last_layer, retain_graph=True)[0]
+        # Get the grad norm for the discriminator loss
+        disc_grads = torch.autograd.grad(fake_loss, last_layer, retain_graph=True)[0]
+        # Calculate the updated discriminator weight based on the grad norms
+        nll_grads_norm = torch.norm(nll_grads)
+        disc_grads_norm = torch.norm(disc_grads)
+        disc_weight = nll_grads_norm / (disc_grads_norm + 1e-4)
+        disc_weight = torch.clamp(disc_weight, 0.0, 1e4).detach()
+        disc_weight *= self.discriminator_weight
+        # Set the discriminator weight. It should be negative to reverse gradients into the autoencoder.
+        self.scale_gradients.set_scale(-disc_weight.item())
+        return disc_weight, nll_grads_norm, disc_grads_norm
+
+    def forward(self, outputs: Dict[str, torch.Tensor], batch: Dict[str, torch.Tensor],
+                last_layer: torch.Tensor) -> Dict[str, torch.Tensor]:
+        losses = {}
+        # Basic L1 reconstruction loss
+        ae_loss = F.l1_loss(outputs['x_recon'], batch[self.input_key], reduction='none')
+        # Count the number of output elements to normalize the loss
+        num_output_elements = ae_loss.numel() // ae_loss.shape[0]
+        losses['ae_loss'] = ae_loss.mean()
+
+        # LPIPs loss. Images for LPIPS must be in [-1, 1]
+        recon_img = outputs['x_recon'].clamp(-1, 1)
+        target_img = batch[self.input_key].clamp(-1, 1)
+        lpips_loss = self.lpips(recon_img, target_img)
+        losses['lpips_loss'] = lpips_loss.mean()
+
+        # Make the nll loss
+        rec_loss = ae_loss + self.lpips_weight * lpips_loss
+        # Note: the +2 here comes from the nll of the laplace distribution.
+        # It's only here to make you feel better by keeping the loss positive for longer.
```

Comment:
The comment is probably not worthwhile, but the factor of 2 is helpful to avoid confusion with negative loss values during training. So I simplified the comment but left the factor of 2.